hindsight/io_modules/derived_stream.lua

```diff
@@ -0,0 +1,377 @@
+2) type - http://docs.aws.amazon.com/redshift/latest/dg/c_Supported_data_types.html
+3) length - Maximum length for string fields (nil for everything else)
+4) attributes - http://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_TABLE_NEW.html
+5) field /function - If this is a string the data is retrieved from read_message(field) otherwise the provided
+   function is invoked and its return value is used for the column data.
+
+-- Hindsight Configuration Examples
+
+format      = "redshift.sql"
+buffer_path = "/mnt/output" -- path where the temporary buffer files are stored
+buffer_size = 10000 * 1000  -- size of the largest buffer before performing a multi-line insert max 16MB - 8KB
+ts_field    = "Timestamp"   -- default
+
+db_config = {
+    host = "example.com",
+    port = 5432,
+    keepalives = 1,
+    keepalives_idle = 30,
+    dbname = "pipeline",
+    user = "user",
+    _password = "password",
+}
+
+-- OR
+
+format      = "redshift.psv"
+buffer_path = "/mnt/output"
+buffer_size = 100 * 1024 * 1024 -- max 1GiB
+s3_path     = "s3://test"
+
+-- OR
+
+format = "protobuf"
+output_path = "/mnt/output" -- path where the daily output files are written
+ts_field    = "Timestamp"   -- default
+
+-- OR
+
+format = "tsv"
+output_path = "/mnt/output" -- path where the daily output files are written
+ts_field    = "Timestamp"   -- default
+nil_value   = "NULL"        -- defaults to an empty string
+
+--]]
+
+local M = {}
+local assert    = assert
+local error     = error
+local ipairs    = ipairs
+local pairs     = pairs
+local pcall     = pcall
+local require   = require
+local setfenv   = setfenv
+local tonumber  = tonumber
+local tostring  = tostring
+local type      = type
+
+local read_config               = read_config
+local read_message              = read_message
+
+local io        = require "io"
+local math      = require "math"
+local os        = require "os"
+local string    = require "string"
+local concat    = require "table".concat
+
+setfenv(1, M) -- Remove external access to contain everything in the module
+
+local SEC_IN_DAY = 60 * 60 * 24
+
+function load_schema(name, schema)
+    local cfg_name          = read_config("cfg_name")
+    local format            = read_config("format")
+    local ts_field          = read_config("ts_field") or "Timestamp"
+    local files             = {} -- manages the derived stream buffer/output files
+
+    if format == "redshift.sql" then
+        local driver    = require "luasql.postgres"
+        local rsql      = require "derived_stream.redshift.sql"
+
+        local function build_connection_string()
+            local t = read_config("db_config")
+            if type(t) ~= "table" then error("db_config must be a table") end
+
+            local options = {}
+            for k,v in pairs(t) do
+                if type(k) ~= "string" then error("invalid connection string key") end
+                if string.match(k, "^_") then
```

Comment:
Variables prefixed with underscore are consider private (i.e., when we display them on a UI),  So I use "_password" in the cfg and correct it for the redshift table in the code.